File: src/main/java/j/command/command/MacroCommand.java
package j.command.command;

import java.util.ArrayDeque;
import java.util.Deque;

public final class MacroCommand implements Command {
  private final Deque<Command> commands = new ArrayDeque<>();

  @Override
  public void execute() {
    for (Command cmd : commands) {
      cmd.execute();
    }
  }

  public void append(Command cmd) {
    if (cmd == this) {
      throw new IllegalArgumentException("infinite loop caused by append");
    }
    commands.push(cmd);
  }

  public void undo() {
    if (!commands.isEmpty()) {
      commands.pop();
    }
  }

  public void clear() {
    commands.clear();
  }
}

File: src/main/kotlin/k/command/command/MacroCommand.kt
package k.command.command

import java.util.*

class MacroCommand : Command {
  private val commands: Deque<Command> = ArrayDeque()

  override fun execute() {
    for (cmd in commands) {
      cmd.execute()
    }
  }

  fun append(cmd: Command) {
    require(!(cmd === this)) { "infinite loop caused by append" }
    commands.push(cmd)
  }

  fun undo() {
    if (!commands.isEmpty()) {
      commands.pop()
    }
  }

  fun clear() {
    commands.clear()
  }
}
